bootstrap/updater.go

```diff
@@ -0,0 +1,407 @@
+)
+
+const (
+	DefaultURL = "http://localhost:3000/bootstrap"
+	DirName    = "bootstrap"
+
+	httpTimeout   = 5 * time.Second
+	notifyTimeout = time.Second
+	schemaFile    = "schema.json"
+	format        = "2006-01-02T15-04-05"
+)
+
+var (
+	ErrEpochOutOfOrder = errors.New("epoch out of order")
+	ErrWrongVersion    = errors.New("wrong schema version")
+	ErrInvalidBeacon   = errors.New("invalid beacon")
+)
+
+type Config struct {
+	URL     string `mapstructure:"bootstrap-uri"`
+	Version string `mapstructure:"bootstrap-version"`
+
+	DataDir   string
+	Interval  time.Duration
+	NumToKeep int
+}
+
+func DefaultConfig() Config {
+	return Config{
+		URL:       DefaultURL,
+		Version:   "https://spacemesh.io/bootstrap.schema.json.1.0",
+		DataDir:   os.TempDir(),
+		Interval:  30 * time.Second,
+		NumToKeep: 10,
+	}
+}
+
+type Updater struct {
+	cfg    Config
+	logger log.Log
+	fs     afero.Fs
+	client *http.Client
+	once   sync.Once
+	eg     errgroup.Group
+
+	mu          sync.Mutex
+	subscribers []chan *VerifiedUpdate
+	latest      *VerifiedUpdate
+}
+
+type Opt func(*Updater)
+
+func WithConfig(cfg Config) Opt {
+	return func(u *Updater) {
+		u.cfg = cfg
+	}
+}
+
+func WithLogger(logger log.Log) Opt {
+	return func(u *Updater) {
+		u.logger = logger
+	}
+}
+
+func WithFilesystem(fs afero.Fs) Opt {
+	return func(u *Updater) {
+		u.fs = fs
+	}
+}
+
+func WithHttpClient(c *http.Client) Opt {
+	return func(u *Updater) {
+		u.client = c
+	}
+}
+
+func New(opts ...Opt) *Updater {
+	u := &Updater{
+		cfg:    DefaultConfig(),
+		logger: log.NewNop(),
+		fs:     afero.NewOsFs(),
+		client: &http.Client{},
+	}
+	for _, opt := range opts {
+		opt(u)
+	}
+	return u
+}
+
+func (u *Updater) Subscribe() chan *VerifiedUpdate {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+	ch := make(chan *VerifiedUpdate, 10)
+	u.subscribers = append(u.subscribers, ch)
+	return ch
+}
+
+func (u *Updater) Load(ctx context.Context) error {
+	verified, err := load(u.fs, u.cfg)
+	if err != nil {
+		return err
+	}
+	if verified != nil {
+		if err = u.updateAndNotify(ctx, verified); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (u *Updater) Start(ctx context.Context) {
+	u.once.Do(func() {
+		u.eg.Go(func() error {
+			if err := u.Load(ctx); err != nil {
+				return err
+			}
+			wait := time.Duration(0)
+			for {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case <-time.After(wait):
+					ctx := log.WithNewSessionID(ctx)
+					if err := u.DoIt(ctx); err != nil {
+						u.logger.With().Error("failed to get bootstrap update", log.Err(err))
+					}
+				}
+				wait = u.cfg.Interval
+			}
+		})
+	})
+}
+
+func (u *Updater) Close() {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+	for _, ch := range u.subscribers {
+		close(ch)
+	}
+	_ = u.eg.Wait()
+}
+
+func (u *Updater) latestUpdateId() uint32 {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+	if u.latest != nil {
+		return u.latest.ID
+	}
+	return 0
+}
+
+func (u *Updater) DoIt(ctx context.Context) error {
```

Comment:
Should this be exported (also `Load`)? From what I understand about the updater it seems like the public API should be:
```
Start
Subscribe
Close
```

I see it's used in the test but the test could be moved to this package.

Reply:
it's a deliberate act on my part. in my view it is more important to put tests in XXX_test pkg here.
there are so many places in go-spacemesh where tests access private members directly and cause race issues.

`Load` and `DoIt` makes testing easier. i'll keep this in mind in my next iteration (there are prolly 2 more PRs) and see how i can reduce the exported methods.

Reply:
yes your preference is clear to me.

if i were developing low level libraries, i'd agree with you. and even there, you will get disagreement across the team.
my consideration is that we are trying to hit genesis testnet next week. so there is time crunch. i don't feel spending more time perfecting this PR is worthwhile.